designs/SearchCommandSpec.md

```diff
@@ -0,0 +1,103 @@
+--------------------
+> Microsoft.Extensions.Logging.Configuration | 5.0.0-preview.6.20305.6 | Downloads: 63,205,816
+```
+
+If a particular source returns no results for a query, it will display output that indicates this.
+
+```
+====================
+Source: dotnet-msbuild
+--------------------
+No results found.
+```
+
+### Optional Arguments
+
+Search provides users with some optional arguments that they can use to customize their search queries:
+
+| Name | Description | Usage |
+| ---  |     ---     |  :-:  |
+| PackageTypes | The package type to use to filter packages (if supported by the source). If the provided package type is not a valid package type as defined by the [Package Type document](https://github.com/NuGet/Home/wiki/Package-Type-%5BPacking%5D), an empty result will returned. | -PackageType `<Package Type>`|
+| PreRelease | Pre-release packages are not included by default, but can be included by using this argument | -PreRelease |
+| Source | Specific package source(s) to search instead of querying the default sources in __nuget.config__ | -Source `<Source URL>`|
+| Take | The number of results to return. The default value is 20. | -Take `<positive integer>` |
+| Verbosity | The level of detail to display in the output. The default is _normal_. (See the note below)  | -Verbosity `<quiet\|normal\|detailed>` |
+| Help | Displays help information for the command | -Help |
+
+__NOTE__
+
+Verbosity Levels:
+
+* _quiet_ - Package ID, Version
+* _normal_ - Package ID, Version, Downloads, Preview of Description
+* _detailed_ - Package ID, Version, Downloads, Full Description, Other information such as the query URL, and the arguments and filters applied to the search
+
+## Future Work
+
+* Integrating the Search feature into dotnet.exe
+* Adding further optional arguments that allow users to customize their search queries
+* Using columns rather than `|` to display results
+* Consider using pagination to display the search results
+* Allow multiple/no query terms
+
+## Open Questions
+
+* What represents the most effective design choice for displaying the search results, in terms of spacing, demarcation, truncating package descriptions, and so on?
+* Would pagination (using the 'more' or 'less' commands) provide a better viewing experience for users? Would users prefer to pipe the search results to these commands themselves, or have Search do it for them? See _Considerations_ for more. 
+* What, if any, additional information should be included for the packages returned by Search?
+
+## Considerations
+
+### 1. Combining the results from different sources
+
+Initially, we considered consolidating all search results from the various sources and interleaving them together, rather than separating them based on their source. However, users may wish to know where a particular package is located. Also, combining the search results for different sources leads to the complex problem of how to combine these results most effectively.
```

Comment:
Don't we combine results in Visual Studio? I imagine we should be able to use the same method to combine results here. My instinct is that I like has results from each source displayed separately, but I wonder how often customers may be using some configuration their team set up and actually aren't aware what source a particular package they need is coming from. @loic-sharma any input here?

Reply:
We are using Lucene.NET.dll to combine the search results. 
We cannot use Lucene.NET in dotnet.exe because of source-build concerns. See https://github.com/dotnet/source-build#net-core-source-build. 

From their repo: 

```
The key goal of this repository is to satisfy the official packaging rules of commonly used Linux distributions, such as Fedora and Debian. Many Linux distributions have similar rules. These rules tend to have two main principles: consistent reproducibility, and source code for everything.

```

Reply:
Also merging search results falls apart when one ft is very slow and another one is fast on the command line. For the VSUI you can adjust the ordering dynamically as slow results come in. For a command line once you write a search result it (can't) be easily undone so you need to wait for all search pages to return before showing anything.